megamolbart/megamolbart/inference_perceiver.py

```diff
@@ -0,0 +1,619 @@
+
+        tokens = self.tokenizer.tokenize(smiles, pad=True)
+
+        # # Append to tokens and mask if appropriate. No Padding for Perceiver
+        if pad_length and self.encoder_type != 'perceiver':
+            for i in range(len(tokens['original_tokens'])):
+                n_pad = pad_length - len(tokens['original_tokens'][i])
+                tokens['original_tokens'][i] += [self.tokenizer.pad_token] * n_pad
+                tokens['masked_pad_masks'][i] += [1] * n_pad
+
+        token_ids = torch.tensor(self.tokenizer.convert_tokens_to_ids(tokens['original_tokens'])).cuda().T
+        pad_mask = torch.tensor(tokens['masked_pad_masks']).bool().cuda().T
+        encode_input = {"encoder_input": token_ids, "encoder_pad_mask": pad_mask}
+
+        embedding = self.model.model.encode(encode_input)
+        torch.cuda.empty_cache()
+        return embedding, pad_mask
+
+    def _inverse_transform_batch(self, memory, mem_pad_mask, k=1, sanitize=True):
+        mem_pad_mask = mem_pad_mask.clone()
+        smiles_interp_list = []
+        (_, batch_size, _) = tuple(memory.size())
+        if self.encoder_type == "perceiver":
+            mem_pad_mask = torch.zeros((memory.shape[0:2]), dtype=mem_pad_mask.dtype, device=mem_pad_mask.device)
+        with torch.no_grad():
+            decode_fn = partial(self.model.model._decode_fn,
+                                mem_pad_mask=mem_pad_mask.type(torch.LongTensor).cuda(),
+                                memory=memory)
+
+            mol_strs, _ = self.model.sampler.beam_decode(decode_fn,
+                                                        batch_size=batch_size,
+                                                        device='cuda',
+                                                        k=k)
+            
+            mol_strs = sum(mol_strs, [])  # flatten list
+            g_smiles = None
+            for smiles in mol_strs:
+                g_smiles = smiles
+                if sanitize:
+                    mol = Chem.MolFromSmiles(g_smiles, sanitize=sanitize)
+                    if mol:
+                        g_smiles = Chem.MolToSmiles(mol)
+                smiles_interp_list.append(g_smiles)
+        return smiles_interp_list
+    
+    # Added this function for easier local working with the model
+    # TODO as the inverse_transform is used via the sql query for generation this function should take in 
+    # either any sized embedding or a list of embeddings and process them via an appropriate batch size
+    # z = model.smiles2embedding(smiles) out_smiles = model.embedding2smiles(z)
+    def embedding2smiles(self, embeddings, mem_pad_mask, k=1, sanitize=True):
+        return self.inverse_transform([embeddings], mem_pad_mask, k, sanitize)[0]
+
+    def inverse_transform(self, embeddings, mem_pad_mask, k=1, sanitize=True):
+        smiles_interp_list = []
+        with torch.no_grad():
+            with concurrent.futures.ThreadPoolExecutor(max_workers=4) as executor:
```

Comment:
Why not doing batch decoding followed by parallel string parsing? 
Running multiple decoding in parallel might slow things down. Did you happen to test the effect on speed?